### PR TITLE
[ESP32] Add API to create user defined size of endpoint array

### PR DESCRIPTION
**Pull request by jadhavrohit924**
- Problem 

> endpoint array size was fixed.

- Changes

> Add API to create user defined size of endpoint array.
> Moved static DRAM data to Heap

- Testing 

>  Tested commissioning and cluster control on esp32.

### CODE REVIEW

File: examples/platform/esp32/mode-support/static-supported-modes-manager.cpp [hunk 1]
```diff
@@ -31,13 +31,15 @@ using SemanticTag          = Structs::SemanticTagStruct::Type;
...
 
-SupportedModesManager::ModeOptionsProvider StaticSupportedModesManager::epModeOptionsProviderList[FIXED_ENDPOINT_COUNT];
+const StaticSupportedModesManager StaticSupportedModesManager::instance = StaticSupportedModesManager();
 
-void StaticSupportedModesManager::InitEndpointArray()
+void StaticSupportedModesManager::InitEndpointArray(int size)
 {
-    for (int i = 0; i < FIXED_ENDPOINT_COUNT; i++)
+    mSize                     = size;
+    epModeOptionsProviderList = new SupportedModesManager::ModeOptionsProvider[mSize];
```
Comment (bzbarsky-apple) [inline examples/platform/esp32/mode-support/static-supported-modes-manager.cpp:41]: Should this be deleted somewhere?  

Is InitEndpointArray() guaranteed to be called only once?

Comment (jadhavrohit924) [reply examples/platform/esp32/mode-support/static-supported-modes-manager.cpp:41]: > Should this be deleted somewhere?

We are deleting it in the destructor.

> Is InitEndpointArray() guaranteed to be called only once?

Yes, InitEndpointArray() should be called only once in the application.

Comment (bzbarsky-apple) [reply examples/platform/esp32/mode-support/static-supported-modes-manager.cpp:41]: > We are deleting it in the destructor.

Where?

Comment (bzbarsky-apple) [reply examples/platform/esp32/mode-support/static-supported-modes-manager.cpp:41]: > Yes, InitEndpointArray() should be called only once in the application.

That is very much not documented.

Comment (jadhavrohit924) [reply examples/platform/esp32/mode-support/static-supported-modes-manager.cpp:41]: > > We are deleting it in the destructor.
> 
> Where?

If you check `~StaticSupportedModesManager()`, we are calling `FreeSupportedModes` for every allocated endpoint

Comment (jadhavrohit924) [reply examples/platform/esp32/mode-support/static-supported-modes-manager.cpp:41]: > > Yes, InitEndpointArray() should be called only once in the application.
> 
> That is very much not documented.

Ok. I'll add comment for this.

Comment (bzbarsky-apple) [reply examples/platform/esp32/mode-support/static-supported-modes-manager.cpp:41]: > we are calling FreeSupportedModes

Where does that delete the array that InitEndpointArray allocates?

Comment (jadhavrohit924) [reply examples/platform/esp32/mode-support/static-supported-modes-manager.cpp:48]: Memory allocated in InitEndpointArray will be needed for lifetime of the program.

Comment (bzbarsky-apple) [reply examples/platform/esp32/mode-support/static-supported-modes-manager.cpp:48]: Then we should document, the reasons we are not freeing it anywhere.

Comment (jadhavrohit924) [reply examples/platform/esp32/mode-support/static-supported-modes-manager.cpp:48]: Added comment [here](https://github.com/project-chip/connectedhomeip/pull/27341/files#diff-a4009156ee3afaf0eab155be1527536b3c7af1992ffbad3bbd113b8d3d0143a3R43)

File: examples/platform/esp32/mode-support/static-supported-modes-manager.h [hunk 2]
```diff
@@ -31,28 +31,40 @@ class StaticSupportedModesManager : public chip::app::Clusters::ModeSelect::Supp
...
+    static ModeOptionsProvider * epModeOptionsProviderList;
 
     void FreeSupportedModes(EndpointId endpointId) const;
 
-public:
     static const StaticSupportedModesManager instance;
 
+public:
+    // InitEndpointArray should be called only once in the application. Memory allocated to the
+    // epModeOptionsProviderList will be needed till the lifetime of the program.
```
Comment (bzbarsky-apple) [inline examples/platform/esp32/mode-support/static-supported-modes-manager.h:44]:
```suggestion
    // epModeOptionsProviderList will be needed for the lifetime of the program, so it's never deallocated.
```

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**